source/jormungandr/jormungandr/equipments/sytral.py

```diff
@@ -165,25 +165,28 @@ def _process_for_equipment_reports(self, data, stop_area_equipments_list):
         :param data: equipments data received from the webservice
         :param stop_area_equipments_list: list of stop_area_equipments from the protobuf response
         """
+
         for sae in stop_area_equipments_list:
-            for st in sae.stop_area.stop_points:
-                for code in st.codes:
-                    if code.type in self.code_types:
-                        equipments_list = jmespath.search(
-                            "equipments_details[?id=='{}']".format(code.value), data
-                        )
+            """
+            Codes might be duplicated across different stop points.
+            Because we report equipments on a stop area basis, we  don't want them duplicated
+            """
+            unique_codes = {str(code): code for st in sae.stop_area.stop_points for code in st.codes}
```

Comment:
Shouldn't you stick to unicode here?
Also, can't you just use a set? (or maybe store stop_points when we will want to display them :-/ )

Not sure: is the type+value couple managed here? Or is it supposed to be (as test's comment makes me think that)?

Reply:
A set can't be used because the item 'Code' is not hashable... However, why create a dict with key = value?

Reply:
I've updated the structure using unicodes, although this string is only used internally to identify an object. Thus, no harm done to external code that may be using unicodes char.

As @mbounabi said, Code is not `hashable`, hence the usage of a dictionary to mimic the behavior of a `set` and I'm using `key=value` because you cannot create a dict without a key. 

The usage of a dictionary was the easiest way for me to filter out identical codes without using a `set`. I'm happy to take any suggestion to do so in a more pythonic way :snake:

Reply:
@pbougue : the couple `type+value` is managed as the `str` or `unicode` function from a protobuf object serialises all members of this object. Therefore, `unicode(code)` will look like `u'type:"TCL_ESCALIER" value: "2"'`

Reply:
I have made a few attempts to improve this part (with a list and append if not in list)... But I'm not pythonically satisfied with everything I could come up with... So, I'll validate this :sweat_smile: